pkg/utils/nodeclaim/nodeclaim.go

```diff
@@ -232,3 +205,39 @@ func UpdateNodeOwnerReferences(nodeClaim *v1beta1.NodeClaim, node *v1.Node) *v1.
 	})
 	return node
 }
+
+// TerminationHandler is a helper function that takes a v1beta1.NodeClaim and calls cloudProvider.Delete() if status condition
+// on nodeClaim is not terminating. If it is terminating then it will check if the kubelet process is still running. It will
+// return an error and a boolean that indicates if the reconciliation should be re-enqueued.
+func TerminationHandler(ctx context.Context, c client.Client, nodeClaim *v1beta1.NodeClaim, node *v1.Node, cloudProvider cloudprovider.CloudProvider) (bool, error) {
+	// Check if the status condition on nodeClaim is Terminating
+	terminationStatus := nodeClaim.StatusConditions().GetCondition(v1beta1.Terminating)
+	if terminationStatus.IsUnknown() || terminationStatus.IsFalse() {
+		// If not then call Delete on cloudProvider to trigger termination and always requeue reconciliation
+		if err := cloudProvider.Delete(ctx, nodeClaim); cloudprovider.IgnoreNodeClaimNotFoundError(err) != nil {
+			return true, fmt.Errorf("terminating cloudprovider instance, %w", err)
+		}
+		return true, SetTerminatingStatus(ctx, c, nodeClaim)
+	}
+	// We do a check on the Ready condition of the node to see if the kubelet process is still running
+	return nodeutils.GetCondition(node, v1.NodeReady).Status == v1.ConditionTrue, nil
+}
+
+// SetTerminatingStatus updates status condition on nodeClaim as Terminating
+func SetTerminatingStatus(ctx context.Context, c client.Client, nodeClaim *v1beta1.NodeClaim) error {
```

Comment:
Added it as another helper function because I considered adding `Terminating` status even if the underlying instance is terminated. In that case we don't need to go through the rest of the flow where we call delete and then set the status. If we don't go ahead with that approach then we can definitely have it inline.

Reply:
Moved the function inline.